alerts/config.go

```diff
@@ -5,116 +5,179 @@ package alerts
+	"github.com/tidepool-org/platform/structure/validator"
 	"github.com/tidepool-org/platform/user"
 )
 
-// Config models a user's desired alerts.
+// Config wraps Alerts to include user relationships.
+//
+// As a wrapper type, Config provides a clear demarcation of what a user
+// controls (Alerts) and what is set by the service (the other values in
+// Config).
 type Config struct {
-	// UserID receives the alerts, and owns this Config.
+	// UserID receives the configured alerts and owns this Config.
 	UserID string `json:"userId" bson:"userId"`
+
 	// FollowedID is the user whose data generates alerts, and has granted
 	// UserID permission to that data.
-	FollowedID      string                 `json:"followedId" bson:"followedId"`
-	UrgentLow       *WithThreshold         `json:"urgentLow,omitempty" bson:"urgentLow,omitempty"`
-	Low             *WithDelayAndThreshold `json:"low,omitempty" bson:"low,omitempty"`
-	High            *WithDelayAndThreshold `json:"high,omitempty" bson:"high,omitempty"`
-	NotLooping      *WithDelay             `json:"notLooping,omitempty" bson:"notLooping,omitempty"`
-	NoCommunication *WithDelay             `json:"noCommunication,omitempty" bson:"noCommunication,omitempty"`
+	FollowedID string `json:"followedId" bson:"followedId"`
+
+	Alerts `bson:",inline,omitempty"`
+}
+
+// Alerts models a user's desired alerts.
+type Alerts struct {
+	UrgentLow       *UrgentLowAlert       `json:"urgentLow,omitempty" bson:"urgentLow,omitempty"`
+	Low             *LowAlert             `json:"low,omitempty" bson:"low,omitempty"`
+	High            *HighAlert            `json:"high,omitempty" bson:"high,omitempty"`
+	NotLooping      *NotLoopingAlert      `json:"notLooping,omitempty" bson:"notLooping,omitempty"`
+	NoCommunication *NoCommunicationAlert `json:"noCommunication,omitempty" bson:"noCommunication,omitempty"`
 }
 
 func (c Config) Validate(validator structure.Validator) {
 	validator.String("UserID", &c.UserID).Using(user.IDValidator)
 	validator.String("FollowedID", &c.FollowedID).Using(user.IDValidator)
-	if c.Low != nil {
-		c.Low.Validate(validator)
+	c.Alerts.Validate(validator)
+}
+
+func (i Alerts) Validate(validator structure.Validator) {
+	if i.Low != nil {
+		i.Low.Validate(validator)
 	}
-	if c.UrgentLow != nil {
-		c.UrgentLow.Validate(validator)
+	if i.UrgentLow != nil {
+		i.UrgentLow.Validate(validator)
 	}
-	if c.High != nil {
-		c.High.Validate(validator)
+	if i.High != nil {
+		i.High.Validate(validator)
 	}
-	if c.NotLooping != nil {
-		c.NotLooping.Validate(validator)
+	if i.NotLooping != nil {
+		i.NotLooping.Validate(validator)
 	}
-	if c.NoCommunication != nil {
-		c.NoCommunication.Validate(validator)
+	if i.NoCommunication != nil {
+		i.NoCommunication.Validate(validator)
 	}
 }
 
 // Base describes the minimum specifics of a desired alert.
 type Base struct {
 	// Enabled controls whether notifications should be sent for this alert.
-	Enabled bool
+	Enabled bool `json:"enabled"`
```

Comment:
It is, however the defaults are just fine. Adding a tag would seem to make the code unnecessarily more complicated/harder to read, IMO. This isn't a strongly held opinion on my part though, so if it still bugs you, let me know, and I'll add the tag.

Reply:
In this past, I used the struct tags as an indicator of where it can be serialized to. So, if the BSON tags were present it was a message that my future self to remember that this struct ends up serialized in BSON.

While it's probably obvious, many of my suggestions are based upon habits I built up over time to help out "future" me have an easier time picking up old code again and preventing future mistakes.

Reply:
> In this past, I used the struct tags as an indicator of where it can be serialized to. So, if the BSON tags were present it was a message that my future self to remember that this struct ends up serialized in BSON.

OK. I'll add them. FWIW, my previous self screams never trust that, probably for reasons similar too, but opposite of yours. :)

Reply:
Yea, I hear you. I suppose if we aren't doing it consistently then it is less-than-helpful.